### PR TITLE
[Form] Added option type checks in some FormTypes

### PR DESCRIPTION
**Pull request by kix**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | none |
| License | MIT |
| Doc PR | none |

Follow-up to #11696. Some of the built-in form types allowed passing invalid options, which lead to errors that were not so easy to understand. This PR adds needed checks where applicable.

Allowed types in `TimeType`'s `empty_value` option were taken from the documentation.

### CODE REVIEW

File: src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php [hunk 1]
```diff
@@ -239,7 +239,9 @@ public function setDefaultOptions(OptionsResolverInterface $resolver)
         ));
 
         $resolver->setAllowedTypes(array(
+            'choices' => array('null', 'array'),
```
Comment (stof) [inline src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php]: The OptionsResolver cannot make an option required only when the other is not there.

**Comment by stof** [comment]
Once the validation for the choices option of the ChoiceType is fixed to avoid breaking BC in child types, I think this should be merged (and it should go in older branches than master)

**Comment by stof** [comment]
@kix can you update the PR to fix the case of the ChoiceType ?

**Comment by kix** [comment]
@stof, yeah, sorry for the wait, been a bit too busy by the end of the year.

**Comment by kix** [comment]
I've had to remove type hints in some `ChoiceList`'s for the test suite to pass. The check is implemented in `ChoiceList::__construct()` instead. Is this wrong?

File: src/Symfony/Component/Form/Extension/Core/ChoiceList/ChoiceList.php [hunk 2]
```diff
@@ -100,7 +108,7 @@ public function __construct($choices, array $labels, array $preferredChoices = a
      * @param array              $labels           The labels belonging to the choices.
      * @param array              $preferredChoices The choices to display with priority.
      */
-    protected function initialize($choices, array $labels, array $preferredChoices)
+    protected function initialize($choices, $labels, $preferredChoices)
```
Comment (stof) [inline src/Symfony/Component/Form/Extension/Core/ChoiceList/ChoiceList.php]: changing this signature is a BC break

Comment (webmozart) [inline src/Symfony/Component/Form/Extension/Core/ChoiceList/ChoiceList.php]: @stof is right. However, this class is not an API class, so technically this change is possible, as long as the change and the upgrade path is documented in the UPGRADE file (as per [our BC policy](http://symfony.com/doc/current/contributing/code/bc.html#changing-classes)). Can you add that information to the UPGRADE file please?

**Comment by webmozart** [comment]
Looks good! Can you please rebase and document how to upgrade subclasses of ChoiceList in the UPGRADE file? After that, I'm :+1:.

File: src/Symfony/Component/Form/Extension/Core/ChoiceList/ChoiceList.php (hunk 2 above)
Comment (kix) [inline src/Symfony/Component/Form/Extension/Core/ChoiceList/ChoiceList.php]: @webmozart, sure, will do.

**Comment by kix** [comment]
@webmozart, does 7d623ab look good to you? Anything I should add or remove?
On build failure: investigating.

File: UPGRADE-3.0.md [hunk 3]
```diff
@@ -110,6 +110,9 @@ UPGRADE FROM 2.x to 3.0
 
 ### Form
 
+ * Type hints for `ChoiceList::initialize()` method's `$labels` and `$preferredChoices`
```
Comment (webmozart) [inline UPGRADE-3.0.md]: "The array type hints..."

File: UPGRADE-3.0.md [hunk 4]
```diff
@@ -110,6 +110,9 @@ UPGRADE FROM 2.x to 3.0
 
 ### Form
 
+ * Type hints for `ChoiceList::initialize()` method's `$labels` and `$preferredChoices` 
+   parameters were removed.
```
Comment (webmozart) [inline UPGRADE-3.0.md]: This should be documented in UPGRADE-2.8.md instead. Also, please add Before/After examples, like everywhere else in the UPGRADE files.

File: src/Symfony/Component/Form/Extension/Core/ChoiceList/SimpleChoiceList.php [hunk 5]
```diff
@@ -11,8 +11,6 @@
 
 namespace Symfony\Component\Form\Extension\Core\ChoiceList;
 
-@trigger_error('The '.__NAMESPACE__.'\SimpleChoiceList class is deprecated since version 2.7 and will be removed in 3.0. Use Symfony\Component\Form\ChoiceList\ArrayChoiceList instead.', E_USER_DEPRECATED);
```
Comment (webmozart) [inline src/Symfony/Component/Form/Extension/Core/ChoiceList/SimpleChoiceList.php]: Why did you remove this line?

**Comment by webmozart** [comment]
Could you please rebase your PR onto 2.8? You need to open a new PR for the 2.8 branch as well.

**Comment by kix** [comment]
@webmozart, sorry, I must've rebased better. Will fix and write a comment once it looks good, OK?

File: UPGRADE-3.0.md (hunk 4 above)
Comment (kix) [inline UPGRADE-3.0.md]: @webmozart, OK, got it, thanks!

File: src/Symfony/Component/Form/Extension/Core/ChoiceList/ChoiceList.php (hunk 2 above)
Comment (fabpot) [inline src/Symfony/Component/Form/Extension/Core/ChoiceList/ChoiceList.php]: Can you add a note in the UPGRADE file, it looks like this is the only remaining thing to do before merging.

**Comment by fabpot** [comment]
@kix This ones looks easy to finish :)

**Comment by kix** [comment]
@fabpot, sure!
2 марта 2016 г. 8:36 PM пользователь "Fabien Potencier" <
notifications@github.com> написал:

File: src/Symfony/Component/Form/Extension/Core/ChoiceList/ChoiceList.php (hunk 2 above)
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/ChoiceList/ChoiceList.php]: @fabpot This file has been removed in 3.0.

This PR should be rebased before a new review IMHO.

Comment (fabpot) [inline src/Symfony/Component/Form/Extension/Core/ChoiceList/ChoiceList.php]: A new PR on master is indeed needed.

Comment (kix) [inline src/Symfony/Component/Form/Extension/Core/ChoiceList/ChoiceList.php]: OK, will try to fix it up.

**Comment by xabbuh** [comment]
@kix Any news here? :)

**Comment by kix** [comment]
@xabbuh, sorry for the wait, I'll try to finish this by the end of the week (hopefully, if nothing unusual comes up)

**Comment by HeahDude** [comment]
@xabbuh @kix This PR is deprecated by #14050, those changes are not needed anymore.

Excepted allow type `array` for `entry_options` which might be relevant today. since it's normalized it will actually make php crash on the call to `$value['block_name']` but it's common sense that options for entry is an array, that's probably why no one complained until now.

**Comment by kix** [comment]
@HeahDude, hmm, thanks, I'll take a look at `$value['block_name']`, I think. Is it OK to close this PR?

**Comment by HeahDude** [comment]
Yes :)

**Comment by HeahDude** [comment]
If you do you should do it on 2.8 I think. Throw the right exception should be a bug fix IMO.

**Comment by kix** [comment]
@HeahDude, agreed, thanks :)
